src/ptwt/_util.py

```diff
@@ -186,6 +187,32 @@ def _check_axes_argument(axes: Sequence[int]) -> None:
         raise ValueError("Cant transform the same axis twice.")
 
 
+def _check_same_device(
+    tensor: torch.Tensor, torch_device: torch.device
+) -> torch.Tensor:
+    if torch_device != tensor.device:
+        raise ValueError("coefficients must be on the same device")
+    return tensor
+
+
+def _check_same_dtype(tensor: torch.Tensor, torch_dtype: torch.dtype) -> torch.Tensor:
+    if torch_dtype != tensor.dtype:
+        raise ValueError("coefficients must have the same dtype")
+    return tensor
+
+
+def _check_same_device_dtype(
+    coeffs: Union[list[torch.Tensor], WaveletCoeff2d, WaveletCoeffNd],
+) -> tuple[torch.device, torch.dtype]:
+    c = _check_if_tensor(coeffs[0])
+    torch_device, torch_dtype = c.device, c.dtype
+
+    _map_result(coeffs, partial(_check_same_device, torch_device=torch_device))
```

Comment:
this code would be much easier to understand if you just wrote out an explicit loop

Reply:
Fair point :+1:

Reply:
The snippet runs the check on all tensors in the coeff data structure.
Similar as below, adding a simple loop would effectively require to copy the `_map_result` functionality.
We could duplicate this function once for each coeff data type to make it more explicit. I would however prefer to address this by making this call more understandable

Reply:
I was a big fan of the _map_result function. We are missing context here. The idea comes from https://jax.readthedocs.io/en/latest/_autosummary/jax.tree.map.html#jax.tree.map, which is a very powerful and reusable concept. We are dealing with a tree-like data structure and want to process all its leaves. I think a tree-map is the right call.

Reply:
We inherit Lists of dicts or tuples as coefficient data type from `pywt`.  Having some kind of a tree map function helps us since it will allow us to process the leaves in every case without having to loop through all of these different data types all of the time. I don't think we want the jax dependency just for this one function. So I think it's ok to have our own, especially since it's not user-facing.